### PR TITLE
[SPARK-47864][PYTHON][DOCS] Enhance "Installation" page to cover all installable options

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to enhance "Installation" page to cover all installable options for PySpark pip installation.

### Why are the changes needed?

To provide more detailed options for user-facing documentation.

### Does this PR introduce _any_ user-facing change?

No API changes, but the user-facing documentation will be improved as below:

## Before

Screenshot 2024-04-17 at 4 13 39 PM

## After
Screenshot 2024-04-18 at 1 54 54 PM
Screenshot 2024-04-18 at 1 55 04 PM

### How was this patch tested?

Manually built docs and check visually.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by itholic** [comment]
cc @HyukjinKwon @ueshin @zhengruifeng @xinrong-meng @allisonwang-db

File: python/docs/source/getting_started/install.rst [hunk 1]
```diff
@@ -165,16 +168,92 @@ To install PySpark from source, refer to |building_spark|_.
...
 
-Note that PySpark requires Java 17 or later with ``JAVA_HOME`` properly set and refer to |downloading|_.
+MLlib
+^^^^^
+
+Installable with ``pip install "pyspark[mllib]"``.
+
+========================== ========================= ======================================================================================
+Package                    Supported version         Note
+========================== ========================= ======================================================================================
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:241]: Let's match the number of `=` to the text

File: python/docs/source/getting_started/install.rst [hunk 2]
```diff
@@ -165,16 +168,92 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+Note that PySpark requires Java 17 or later with ``JAVA_HOME`` properly set and refer to |downloading|_.
+
+
+.. _optional-dependencies:
+
+Optional dependencies
+~~~~~~~~~~~~~~~~~~~~~
+
+PySpark has several optional dependencies that enhance its functionality for specific modules. These dependencies are only required for certain features and are not necessary for the basic functionality of PySpark. If these optional dependencies are not installed, PySpark will function correctly for basic operations but will raise an ``ImportError`` when you try to use features that require these dependencies.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:193]: let's add some newlines

File: python/docs/source/getting_started/install.rst (hunk 1 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst]: matched!

File: python/docs/source/getting_started/install.rst (hunk 2 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst]: Added!

**Comment by itholic** [comment]
Add more dependencies and updated screen capture from PR description accordingly.

File: python/docs/source/getting_started/install.rst [hunk 3]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+======= ================= ====================================================================================================
+`numpy` >=1.21            Essential for mathematical operations within MLlib, improves performance and accuracy of algorithms.
+======= ================= ====================================================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **torch**: Utilized for machine learning model training on PySpark.
+- **torchvision**: Supports image and video processing within PySpark models.
+- **torcheval**: Facilitates model evaluation metrics in PySpark.
+- **deepspeed; sys_platform != 'darwin'**: Provides high-performance model training optimizations. Installable on non-Darwin systems.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:267]:
```suggestion
- **deepspeed**: Provides high-performance model training optimizations. Installable on non-Darwin systems.
```

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng and @WeichenXu123 would you mind reviewing this please?

**Review by WeichenXu123 — APPROVED** [review]
LGTM!

File: python/docs/source/getting_started/install.rst [hunk 4]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+======= ================= ====================================================================================================
+Package Supported version Note
+======= ================= ====================================================================================================
+`numpy` >=1.21            Essential for mathematical operations within MLlib, improves performance and accuracy of algorithms.
+======= ================= ====================================================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **torch**: Utilized for machine learning model training on PySpark.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:264]: I think `torch`, `torchvision`, `torcheval` and `deepspeed` are for ML.

File: python/docs/source/getting_started/install.rst [hunk 5]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+`numpy`   >=1.21            Facilitates complex numerical operations within Spark.
+========= ================= =====================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **mlflow**: Enhances machine learning lifecycle management, including experiment tracking and model deployment.
+- **plotly, matplotlib**: Provide advanced plotting capabilities for visualization.
+
+
+ML
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:235]: Let's use the official name `MLLib DataFrame-based API`

File: python/docs/source/getting_started/install.rst [hunk 6]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+Required dependencies
+~~~~~~~~~~~~~~~~~~~~~
+
+PySpark requires the following dependencies.
+
+========================== ========================= ============================================
+Package                    Supported version         Note
+========================== ========================= ============================================
+`py4j`                     >=0.10.9.7                Essential for Python to interface with the
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:180]: Can we make the note simpler? e.g., Used to interact to JVM.

File: python/docs/source/getting_started/install.rst [hunk 7]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+                                                     Java objects in Spark; ensures seamless
+                                                     interaction between Python and JVM.
+========================== ========================= ============================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **memory-profiler**: Useful for diagnosing and analyzing memory usage in PySpark applications.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:187]:
```suggestion
- **memory-profiler**: Used for PySpark UDF memory profiling, ``spark.profile.show(...)`` and ``spark.sql.pyspark.udf.profiler``.
```

cc @xinrong-meng

File: python/docs/source/getting_started/install.rst [hunk 8]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+Spark SQL
+^^^^^^^^^
+
+Installable with ``pip install "pyspark[sql]"``.
+
+========= ================= ==================================================================
+Package   Supported version Note
+========= ================= ==================================================================
+`pandas`  >=1.4.4           Enables seamless DataFrame operations between Spark and Pandas.
+`pyarrow` >=10.0.0          Optimizes data conversion and transfer between PySpark and Pandas.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:211]:
```suggestion
`pyarrow` >=10.0.0          For faster data conversion/transfer between PySpark and pandas.
```

File: python/docs/source/getting_started/install.rst [hunk 9]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+`numpy`   >=1.21            Essential for numerical data manipulation within PySpark.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:212]: Let's remove this. This is really for pandas and PyArrow

File: python/docs/source/getting_started/install.rst [hunk 10]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+Pandas API on Spark
+^^^^^^^^^^^^^^^^^^^
+
+Installable with ``pip install "pyspark[pandas_on_spark]"``.
+
+========= ================= =====================================================================
+Package   Supported version Note
+========= ================= =====================================================================
+`pandas`  >=1.4.4           Required for utilizing the Pandas API features in Spark.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:224]:
```suggestion
`pandas`  >=1.4.4           Required for Pandas API on Spark
```

File: python/docs/source/getting_started/install.rst [hunk 11]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+`pyarrow` >=10.0.0          Ensures efficient data handling and performance in Pandas operations.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:225]: For faster data conversion/transfer between PySpark and pandas.

File: python/docs/source/getting_started/install.rst [hunk 12]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+`numpy`   >=1.21            Facilitates complex numerical operations within Spark.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:226]: ditto. we don't really need this.

File: python/docs/source/getting_started/install.rst [hunk 13]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+========= ================= =====================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **mlflow**: Enhances machine learning lifecycle management, including experiment tracking and model deployment.
+- **plotly, matplotlib**: Provide advanced plotting capabilities for visualization.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:232]: Let's list `plotly` and `matplotlib` separately, and mention which API we use, and say `plotly` is preferred.

File: python/docs/source/getting_started/install.rst [hunk 14]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+Package   Supported version Note
+========= ================= =====================================================================
+`pandas`  >=1.4.4           Required for utilizing the Pandas API features in Spark.
+`pyarrow` >=10.0.0          Ensures efficient data handling and performance in Pandas operations.
+`numpy`   >=1.21            Facilitates complex numerical operations within Spark.
+========= ================= =====================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **mlflow**: Enhances machine learning lifecycle management, including experiment tracking and model deployment.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:231]: Mention which API it uses

File: python/docs/source/getting_started/install.rst [hunk 15]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+ML
+^^
+
+Installable with ``pip install "pyspark[ml]"``.
+
+======= ================= =======================================================================
+Package Supported version Note
+======= ================= =======================================================================
+`numpy` >=1.21            Supports advanced data manipulation and algorithm implementation in ML.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:243]: Just say that it's "Required for MLLib DataFrame-based API"

File: python/docs/source/getting_started/install.rst [hunk 16]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+MLlib
+^^^^^
+
+Installable with ``pip install "pyspark[mllib]"``.
+
+======= ================= ====================================================================================================
+Package Supported version Note
+======= ================= ====================================================================================================
+`numpy` >=1.21            Essential for mathematical operations within MLlib, improves performance and accuracy of algorithms.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:259]: Required for MLLib

File: python/docs/source/getting_started/install.rst [hunk 17]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+======= ================= =======================================================================
+Package Supported version Note
+======= ================= =======================================================================
+`numpy` >=1.21            Supports advanced data manipulation and algorithm implementation in ML.
+======= ================= =======================================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **scipy**: Essential for scientific computing and statistical functions in ML.
+- **scikit-learn**: Required for implementing machine learning algorithms.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:249]: @WeichenXu123 and @zhengruifeng can you help writing better description?

File: python/docs/source/getting_started/install.rst [hunk 18]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+Spark Connect
+^^^^^^^^^^^^^
+
+Installable with ``pip install "pyspark[connect]"``.
+
+========================== ================= ====================================================================
+Package                    Supported version Note
+========================== ================= ====================================================================
+`pandas`                   >=1.4.4           Facilitates DataFrame handling and manipulation.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:278]: Just say that Required for Spark Connect

File: python/docs/source/getting_started/install.rst [hunk 19]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+`pyarrow`                  >=10.0.0          Crucial for data serialization and network communication efficiency.
+`numpy`                    >=1.21            Enhances data processing capabilities.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:280]: this too. no need to mention actually.

**Review by dongjoon-hyun — COMMENTED** [review]
Just a question. Are we going to allow different versions of dependencies per module from PySpark 4.0.0?

File: python/docs/source/getting_started/install.rst [hunk 20]
```diff
@@ -165,16 +168,117 @@ To install PySpark from source, refer to |building_spark|_.
...
+- **torch**: Utilized for machine learning model training on PySpark.
+- **torchvision**: Supports image and video processing within PySpark models.
+- **torcheval**: Facilitates model evaluation metrics in PySpark.
+- **deepspeed; sys_platform != 'darwin'**: Provides high-performance model training optimizations. Installable on non-Darwin systems.
+
+
+Spark Connect
+^^^^^^^^^^^^^
+
+Installable with ``pip install "pyspark[connect]"``.
```
Comment (allisonwang-db) [inline python/docs/source/getting_started/install.rst:273]: Shall we move this section to the top?
Also, what's the difference between installing `pyspark[connect]` and `pyspark-connect`? Should we highlight the difference here, or add a documentation link to the pyspark-connect package?

**Review by allisonwang-db — APPROVED** [review]

File: python/docs/source/getting_started/install.rst (hunk 6 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst:180]: Sure

File: python/docs/source/getting_started/install.rst (hunk 20 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst]: Moved the section to the top.

IIRC `pyspark-connect` is not yet officially supported by pip install so I don't mentioned for now. May can update the documentation when it's officially released. @HyukjinKwon could you double check if I understand correctly?

File: python/docs/source/getting_started/install.rst [hunk 21]
```diff
@@ -165,16 +168,109 @@ To install PySpark from source, refer to |building_spark|_.
...
+
+========================== ================= ====================================================================
+Package                    Supported version Note
+========================== ================= ====================================================================
+`pandas`                   >=1.4.4           Required for Spark Connect.
+`pyarrow`                  >=10.0.0          Crucial for data serialization and network communication efficiency.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:209]:
```suggestion
`pyarrow`                  >=10.0.0          Used for data serialization and network communication efficiency.
```

File: python/docs/source/getting_started/install.rst [hunk 22]
```diff
@@ -165,16 +168,109 @@ To install PySpark from source, refer to |building_spark|_.
...
+`grpcio`                   >=1.62.0          Necessary for implementing RPC functionalities in Spark Connect.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:210]:
```suggestion
`grpcio`                   >=1.62.0          Used for implementing RPC functionalities in Spark Connect.
```

File: python/docs/source/getting_started/install.rst (hunk 21 above)
Comment (HyukjinKwon) [reply python/docs/source/getting_started/install.rst:209]: Can we just say that `Required for Spark Connect` for all the dependencies here? The notes here look a bit inconsistent

File: python/docs/source/getting_started/install.rst [hunk 23]
```diff
@@ -165,16 +168,109 @@ To install PySpark from source, refer to |building_spark|_.
...
+========= ================= ===============================================================
+`pandas`  >=1.4.4           Required for Pandas API on Spark.
+`pyarrow` >=10.0.0          For faster data conversion/transfer between PySpark and pandas.
+========= ================= ===============================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **mlflow**: Enhances machine learning lifecycle management. Used for ``pyspark.pandas.mlflow``.
+- **plotly**: Provide advanced plotting capabilities for visualization.
+- **matplotlib**: Provide plotting, but **plotly** is recommended. The default is **plotly**.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:244]: Can you say `but **plotly** is recommended. The default is **plotly**.` at `**plotly**`?

File: python/docs/source/getting_started/install.rst [hunk 24]
```diff
@@ -165,16 +168,109 @@ To install PySpark from source, refer to |building_spark|_.
...
+========= ================= ===============================================================
+Package   Supported version Note
+========= ================= ===============================================================
+`pandas`  >=1.4.4           Required for Pandas API on Spark.
+`pyarrow` >=10.0.0          For faster data conversion/transfer between PySpark and pandas.
+========= ================= ===============================================================
+
+Additional libraries that enhance functionality but are not included in the installation packages:
+
+- **mlflow**: Enhances machine learning lifecycle management. Used for ``pyspark.pandas.mlflow``.
```
Comment (HyukjinKwon) [inline python/docs/source/getting_started/install.rst:242]: `Enhances machine learning lifecycle management` this sounds funny. refer to https://github.com/mlflow/mlflow

File: python/docs/source/getting_started/install.rst (hunk 20 above)
Comment (HyukjinKwon) [reply python/docs/source/getting_started/install.rst]: I already documented above. We should have the docs before releasing it out.
- `pyspark[connect]` with Jars
- `pyspark-connect` without jars.

File: python/docs/source/getting_started/install.rst (hunk 21 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst:209]: Sounds better. Let me update

File: python/docs/source/getting_started/install.rst (hunk 24 above)
Comment (itholic) [reply python/docs/source/getting_started/install.rst:242]: Let me just simplify the context without verbose explanation.

**Comment by itholic** [comment]
> Just a question. Are we going to allow different versions of dependencies per module from PySpark 4.0.0?

IIRC there are currently no plans, but we may consider it if there is a requirement to use a different version of a specific module.

The purpose of this work is simply to provide users with more explanations of various install options, like [Pandas docs](https://pandas.pydata.org/docs/getting_started/install.html#dependencies).

File: python/docs/source/getting_started/install.rst (hunk 17 above)
Comment (zhengruifeng) [reply python/docs/source/getting_started/install.rst:261]: I think we can use this description first.

**Review by zhengruifeng — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for the answer.

+1, LGTM from my side.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**